pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -658,6 +658,22 @@ func (lrw *logicalReplicationWriterProcessor) flushChunk(
 	for len(chunk) > 0 {
 		batch := chunk[:min(batchSize, len(chunk))]
 		chunk = chunk[len(batch):]
+
+		// Make sure we're not ingesting events with origin TS in the future.
+		hlcNow := lrw.FlowCtx.Cfg.DB.KV().Clock().Now()
+		logClock := true
+		for _, kv := range batch {
+			if ts := kv.KeyValue.Value.Timestamp; ts.After(hlcNow) {
+				if logClock || log.V(1) {
+					log.Warningf(ctx, "event timestamp %s is ahead of local clock %s; delaying batch...", ts, hlcNow)
+					logClock = false
+				}
+				if err := lrw.FlowCtx.Cfg.DB.KV().Clock().SleepUntil(ctx, ts); err != nil {
+					return flushStats{}, err
+				}
```

Comment:
I wonder if we'll want a metric here eventually for how often we waited.

Reply:
Eh, we're logging if it happens which seems like enough for now. That metric will be zero ~all the time and just more tsdb load to maintain.